client/network/src/transport.rs

```diff
@@ -55,17 +55,25 @@ pub fn build_transport(
 	// Build the base layer of the transport.
 	let transport = if !memory_only {
 		let tcp_config = tcp::Config::new().nodelay(true);
-		let desktop_trans = tcp::tokio::Transport::new(tcp_config.clone());
-		let desktop_trans = websocket::WsConfig::new(desktop_trans)
+		let tcp_trans = tcp::tokio::Transport::new(tcp_config.clone());
+		let desktop_trans = websocket::WsConfig::new(tcp_trans)
 			.or_transport(tcp::tokio::Transport::new(tcp_config.clone()));
 		let dns_init = dns::TokioDnsConfig::system(desktop_trans);
 		EitherTransport::Left(if let Ok(dns) = dns_init {
-			EitherTransport::Left(dns)
+			let tcp_trans = tcp::tokio::Transport::new(tcp_config.clone());
+			let desktop_trans = websocket::WsConfig::new(tcp_trans)
+				.or_transport(tcp::tokio::Transport::new(tcp_config.clone()));
+			let dns_init = dns::TokioDnsConfig::system(desktop_trans);
+			// Secure Websocket transport needs unresolved addresses, so we join DNS transport with
+			// yet another instance of Websocket transport.
+			EitherTransport::Left(
+				websocket::WsConfig::new(dns).or_transport(dns_init.expect("same config to work")),
```

Comment:
`dns` here already contains an instance of `WsConfig`. Intuitively, `DnsConfig` required by `WsConfig::new` should only provide DNS and TCP transport. May be it makes sense separating DNS carrying TCP+WS and "bare" DNS for WSS?

Reply:
Also, I don't know internals of libp2p good enough to reason with confidence, but may be `WsConfig` based on `DnsConfig` is capable of handling raw `ip4`/`ip6` addresses and we don't need two instances of WS transport?

Reply:
> May be it makes sense separating DNS carrying TCP+WS and "bare" DNS for WSS?

good point. done in https://github.com/paritytech/substrate/pull/13152/commits/756ef866b2fc8b78c93568078310af4b7ce9907c

Reply:
> Also, I don't know internals of libp2p good enough to reason with confidence, but may be `WsConfig` based on `DnsConfig` is capable of handling raw `ip4`/`ip6` addresses and we don't need two instances of WS transport?

Did you try running without second instance of WS transport for bare IP addresses?

Reply:
> Did you try running without second instance of WS transport for bare IP addresses?

Trying right now. It should work in theory

Reply:
It works 👍